### PR TITLE
Add minimum capability requirement for AWQ

### LINKED ISSUE #1063: AWQ does not support Turing GPUs
@WoosukKwon Doesn't it support Turing arch? my GPU's compute capabitlity is 7.5. CUDA-12.1.

build Error message:

  `ptxas /tmp/tmpxft_0006e7c4_00000000-6_gemm_kernels.ptx, line 928; error   : Feature '.m16n8k16' requires .target sm_80 or higher`

If not, hope can add backward compatibility for kernel build.

_Originally posted by @esmeetu in https://github.com/vllm-project/vllm/issues/1032#issuecomment-1722179620_

### PR DESCRIPTION
**Pull request by WoosukKwon**
Closes #1063

### CODE REVIEW

**Comment by esmeetu** [comment]
@WoosukKwon I think setup.py file might need add compute capability check whether to build quant kernel or not.

**Comment by WoosukKwon** [comment]
@esmeetu Thanks for the suggestion! I've instead added a guard to prevent compilation for the supported GPUs.

File: csrc/quantization/awq/dequantize.cuh [hunk 1]
```diff
@@ -11,9 +11,14 @@ Modified from NVIDIA FasterTransformer: https://github.com/NVIDIA/FasterTransfor
 
 #pragma once
 
+namespace vllm {
+namespace awq {
```
Comment (zhuohan123) [inline csrc/quantization/awq/dequantize.cuh:15]: Why is this namespace needed?

File: vllm/model_executor/model_loader.py [hunk 2]
```diff
@@ -68,6 +68,14 @@ def get_model(model_config: ModelConfig) -> nn.Module:
...
                                         model_config.model,
                                         model_config.download_dir)
+        capability = torch.cuda.get_device_capability()
+        capability = capability[0] * 10 + capability[1]
+        if capability < quant_config.get_min_capability():
+            raise ValueError(
+                f"The quantization method {model_config.quantization} is not "
+                "supported for the current GPU. "
+                f"Minimum capability: {quant_config.get_min_capability()}. "
+                f"Current capability: {capability}.")
```
Comment (zhuohan123) [inline vllm/model_executor/model_loader.py:78]: Why do we need the `assert false` in C++ if we have the check here?

**Review by zhuohan123 — COMMENTED** [review]
Left some comments.

File: vllm/model_executor/model_loader.py (hunk 2 above)
Comment (zhuohan123) [reply vllm/model_executor/model_loader.py:78]: See the comments in the PR and got it.

File: csrc/quantization/awq/dequantize.cuh (hunk 1 above)
Comment (WoosukKwon) [reply csrc/quantization/awq/dequantize.cuh:15]: They are optional, but for better coding convention. From google cpp style guide:

> With few exceptions, place code in a namespace.

Namespace prevents namespace conflicts, so it's pretty useful for external code like the AWQ kernels.

File: vllm/model_executor/model_loader.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/model_executor/model_loader.py:78]: It's problematic when we want to build the wheel for all GPU architectures (e.g., for pypi publication or building docker image). In such a case, we cannot selectively include the extension according to the architecture. Therefore, I believe this is an easier solution, and in fact we already used this kind of guard for bfloat16 attention kernels, which do not support Turing and Volta GPUs.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the fix!

**Pull request merged**